// happner-security/test/lib/0-config.js
var path = require('path');

module.exports = {
  name: 'happner-cloud',
  datalayer: {
    log_level: 'info|error|warning',
    persist:true
  },
  modules: {
    'happner-security-user': {
      construct:{
        name:"UserManager"
      },
      path: path.resolve('././index.js')
    },
    'happner-security-network': {
      construct:{
        name:"NetworkManager"
      },
      path: path.resolve('././index.js')
    },
    'happner-security-group': {
      construct:{
        name:"GroupManager"
      },
      path: path.resolve('././index.js')
    }
  },
  components:{
    'happnerSecurityNetwork': {
      accessLevel:"mesh",
      startMethod: "start",
      moduleName:"happner-security-network",
      schema:{
        "exclusive":false,//means we dont dynamically share anything else
        "methods":{
          "start":{
            type:"async",
            parameters:[
             {"required":true, "value":{/*TODO - configuration goes here*/}}  
            ]
          }
        }
      }
    },
    'happnerSecurityUser': {
      accessLevel:"mesh",
      startMethod: "start",
      moduleName:"happner-security-user",
      schema:{
        "exclusive":false,//means we dont dynamically share anything else
        "methods":{
          "start":{
            type:"async",
            parameters:[
             {"required":true, "value":{/*TODO - configuration goes here*/}}  
            ]
          }
        }
      }
    },
    'happnerSecurityGroup': {
      accessLevel:"mesh",
      startMethod: "start",
      moduleName:"happner-security-group",
      schema:{
        "exclusive":false,//means we dont dynamically share anything else
        "methods":{
          "start":{
            type:"async",
            parameters:[
             {"required":true, "value":{/*TODO - configuration goes here*/}}  
            ]
          }
        }
      }
    }
  }
}
