// happner-security/lib/group.js
module.exports = GroupManager;

function GroupManager() {}

GroupManager.prototype.PATH_GroupManager = '/HAPPNER-DTA/NETWORK/GroupManager'; //permissions changed
GroupManager.prototype.PATH_SD_UPDATE = '/HAPPNER-EVT/NETWORK/SD_UPDATE'; //security directory update

GroupManager.prototype.create = function($happn, params, callback) {

	try{

	}catch(e){
		callback(e);
	}

}

GroupManager.prototype.delete = function($happn, params, callback) {

	try{

	}catch(e){
		callback(e);
	}
  
}

GroupManager.prototype.update = function($happn, params, callback) {

	try{

	}catch(e){
		callback(e);
	}

}

GroupManager.prototype.addPermission = function($happn, params, callback){

	try{

	}catch(e){
		callback(e);
	}

}

GroupManager.prototype.removePermission = function($happn, params, callback){

	try{

	}catch(e){
		callback(e);
	}

}

GroupManager.prototype.clone = function($happn, params, callback){

	try{

	}catch(e){
		callback(e);
	}

}

GroupManager.prototype.__cache = null;
GroupManager.prototype.list = function($happn, params, callback){

	try{

	}catch(e){
		callback(e);
	}

}

GroupManager.prototype.__refreshCache = function($happn, params, callback){

	try{

		

		if (callback) callback();
	}catch(e){
		$happn.log.error('error refreshing GroupManager cache', e);
		if (callback) callback(e);
	}

}

GroupManager.prototype.start = function($happn, params, callback){

	try{

		var _this = this;

		_this.__refreshCache($happn, {}, function(e){

			if (e) return callback(e);

			$happn.event.happnerSecurityNetwork.on('NETWORK-EVT', function(){
				_this.__refreshCache($happn, {});
			}, callback);

		});

	}catch(e){
		callback(e);
	}

	
}
