// happner-security/lib/group.js
module.exports = Group;

function Group() {}

Group.prototype.create = function($happn, params, callback) {

	try{

	}catch(e){
		callback(e);
	}

}

Group.prototype.delete = function($happn, params, callback) {

	try{

	}catch(e){
		callback(e);
	}
  
}

Group.prototype.update = function($happn, params, callback) {

	try{

	}catch(e){
		callback(e);
	}

}

Group.prototype.addPermission = function($happn, params, callback){

	try{

	}catch(e){
		callback(e);
	}

}

Group.prototype.removePermission = function($happn, params, callback){

	try{

	}catch(e){
		callback(e);
	}

}

Group.prototype.clone = function($happn, params, callback){

	try{

	}catch(e){
		callback(e);
	}

}

Group.prototype.__cache = null;
Group.prototype.list = function($happn, params, callback){

	try{

	}catch(e){
		callback(e);
	}

}

Group.prototype.__refreshCache = function($happn, params){

	try{
		
	}catch(e){
		
	}

}

Group.prototype.start = function($happn, params, callback){

	try{

		var _this = this;

		$happn.event.happnerSecurityNetwork.on('NETWORK-EVT', function(){
			_this.__refreshCache($happn, {});
		}, callback);

	}catch(e){
		callback(e);
	}

	
}
