// happner-security/lib/user.js
module.exports = UserManager;

function UserManager() {}

var users = {
	admin: {
		name: 'admin',
		network: '/OEM/company/plant/' + 'happner-cloud',
		userType: 'builtin',
		groups: [

		]
	}
};

UserManager.prototype.create = function($happn, params, callback) {

	var name = params.name;  // Becomes the key. (Using a natural keys approach)
	var user = {
		name: name,
		network: '/OEM/company/plant/' + $happn._mesh.config.name,
		groups: [],
		// methods: [],
		// events: []
	}


	// var response = JSON.parse(JSON.stringify(params.UserManager))
 //  	callback(null, params.UserManager);
}

UserManager.prototype.delete = function($happn, params, callback) {
	var name = params.name; // (KEY) The user to be deleted

  callback(null);
}

UserManager.prototype.update = function($happn, params, callback) {
  callback(null);
}

UserManager.prototype.addGroup = function($happn, params, callback){
	callback(null);
}

UserManager.prototype.removeGroup = function($happn, params, callback){
	callback(null);
}

UserManager.prototype.clone = function($happn, params, callback){
	callback(null);
}

UserManager.prototype.__cache = null;
UserManager.prototype.list = function($happn, params, callback){

	try{

	}catch(e){
		callback(e);
	}

}

UserManager.prototype.__refreshCache = function($happn, params){

	try{
		
	}catch(e){
		
	}

}

UserManager.prototype.start = function($happn, params, callback){

	try{

		var _this = this;

		$happn.event.happnerSecurityNetwork.on('NETWORK-EVT', function(){
			_this.__refreshCache($happn, {});
		}, callback);

	}catch(e){
		callback(e);
	}

	
}
